### DISCUSSION TITLE
How to set 'environment' dynamically?

### DISCUSSION THREAD

**Discussion by vitalykarasik** [community/community#38178] [Actions] [answered]
I want to use 'environment' for updating Jira Deployments dashboard.
In all examples I see that job.environment setting is static, i.e.

`build_deploy:
    environment: staging`

I'm curious if/how I can set 'environment' inside my job. Currently I have one build-and-deploy job, which detect deployment target based on Github branch, i.e. I have something like that:

```
- name: set environment (production) 
  if:  github.ref_name == 'master' 
  run: |
      echo "NODE_ENV=production" >> $GITHUB_ENV  
```

So I'd like to set 'environment' in the same way, without refactoring my workflow.

TIA, Vitaly

**Comment by ChristopherHX** [comment, marked as answer]
> without refactoring my workflow.

Depends on what is refactoring for you, you can't set the environment once your job is already running only the deployment url is allowed to change during a deployment job

```yaml
environment:
  name: ${{ github.ref_name == 'master' && 'production' || 'staging' }}
```

On newer GitHub actions services this should also work
```yaml
environment: ${{ github.ref_name == 'master' && 'production' || 'staging' }}
```

**Reply by vitalykarasik** [reply]
In other words, I should use a one-liner macro.  Ideally, I need some kind if /elseif flow for processing 3 environments, I'll try to write that.
Many thanks, Vitaly.

**Reply by ChristopherHX** [reply]
You don't need a one liner, here an elseif flow in multi line
```yaml
environment: |-
  ${{
     github.ref_name == 'master' && 'production'
  || github.ref_name == 'dev'    && 'dev'
  ||                                'staging'
  }}
```
Make shure that only the last `||` component can be null/0/''.
Note, if you use `|` instead of `|-` you get a newline appended to your expression result.

**Reply by vitalykarasik** [reply]
Wow, fantastic, many thanks!  Somehow I was sure that macro expression has to be oneliner...

**Comment by debajit-altera** [comment]
I am using a re-usable workflow and want to set the environment by the input. It does not throw error, however it is not set either since, the re-usable workflow cant use my environment secret.
On the other hand if I use github.input , its setting dynamically. Can I get clarity for this behavior please? Any documentation is welcome as well.
